### PR TITLE
ZFS support

### PR DESCRIPTION
**Pull request by Jeroen0494**
Hi,

This PR includes basic support for ZFS. Right now it is recommended to run these profiles in complain mode. 

Containerd with the ZFS snapshotter requires some additional configuration.

```
ALLOWED containerd mount /run/containerd/io.containerd.grpc.v1.cri/sandboxes/2f160184804b3f0f84c202e0ba5abaed348ddd14137afae848c909712f050056/shm/ info="failed srcname match" comm=containerd error=-13 fstype=tmpfs srcname=shm flags="rw, nosuid, nodev, noexec"
ALLOWED containerd mount /run/containerd/io.containerd.grpc.v1.cri/sandboxes/68a87340eac18d5b3bfc6d342a4baa291f8426b08adc8692bfe99f705746ef69/shm/ info="failed srcname match" comm=containerd fstype=tmpfs error=-13 srcname=shm flags="rw, nosuid, nodev, noexec"
ALLOWED containerd mount /run/containerd/io.containerd.grpc.v1.cri/sandboxes/5211fa8c45692abded33bd20a4a177d8d2916e94db5602560713ed57440bce55/shm/ info="failed srcname match" comm=containerd fstype=tmpfs srcname=shm flags="rw, nosuid, nodev, noexec" error=-13
ALLOWED containerd open /dev/block/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/bsg/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/bus/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/bus/usb/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/bus/usb/001/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/bus/usb/002/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/char/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/cpu/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/cpu/0/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/cpu/1/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/by-id/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/by-label/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/by-partlabel/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/by-partuuid/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/by-path/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/disk/by-uuid/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/dma_heap/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/dri/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/dri/by-path/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/hugepages/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/input/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/input/by-id/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/input/by-path/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/mapper/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/net/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/rpool/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/snd/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/snd/by-path/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/vfio/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/zvol/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd open /dev/zvol/rpool/ comm=containerd requested_mask=r denied_mask=r
ALLOWED containerd mount /var/lib/containerd/tmpmounts/containerd-mount3231530792/ info="failed srcname match" comm=containerd srcname=rpool/containerd/196 error=-13 fstype=zfs
ALLOWED containerd mount /var/lib/containerd/tmpmounts/containerd-mount1861441902/ info="failed srcname match" comm=containerd error=-13 fstype=zfs srcname=rpool/containerd/197
ALLOWED containerd umount /var/lib/containerd/tmpmounts/containerd-mount3231530792/ comm=containerd
ALLOWED containerd umount /var/lib/containerd/tmpmounts/containerd-mount1861441902/ comm=containerd
ALLOWED containerd mount /var/lib/containerd/tmpmounts/containerd-mount3997796280/ info="failed srcname match" comm=containerd fstype=zfs srcname=rpool/containerd/196 error=-13
ALLOWED containerd umount /var/lib/containerd/tmpmounts/containerd-mount3997796280/ comm=containerd
ALLOWED containerd mount /var/lib/containerd/tmpmounts/containerd-mount813552795/ info="failed srcname match" comm=containerd error=-13 fstype=zfs srcname=rpool/containerd/198
ALLOWED containerd umount /var/lib/containerd/tmpmounts/containerd-mount813552795/ comm=containerd
ALLOWED containerd mount /var/lib/containerd/tmpmounts/containerd-mount3381430269/ info="failed srcname match" comm=containerd error=-13 fstype=zfs srcname=rpool/containerd/197
ALLOWED containerd umount /var/lib/containerd/tmpmounts/containerd-mount3381430269/ comm=containerd
ALLOWED containerd mount /var/lib/containerd/tmpmounts/containerd-mount716833569/ info="failed srcname match" comm=containerd fstype=zfs srcname=rpool/containerd/198 error=-13
ALLOWED containerd umount /var/lib/containerd/tmpmounts/containerd-mount716833569/ comm=containerd
ALLOWED containerd open /proc/2516/attr/current comm=containerd requested_mask=r denied_mask=r
```

I've denied a lot of `/dev/` devices that it shouldn't have access to, I think containerd just traverses the whole `/dev/` directory.
Let me know what you think!

### CODE REVIEW

File: apparmor.d/groups/virt/containerd [hunk 1]
```diff
@@ -40,5 +44,34 @@ profile containerd @{exec_path} {
...
+  # zfs snapshotter
+  /{usr/,}{local/,}{s,}bin/zfs Px,
+  mount fstype=zfs -> /var/lib/containerd/tmpmounts/containerd-mount[0-9]*/,
+  umount -> /var/lib/containerd/tmpmounts/containerd-mount[0-9]*/,
+  /var/lib/containerd/tmpmounts/containerd-mount[0-9]*/lib{64,}/** l,
+  deny /dev/bsg/ r,
+  deny /dev/bus/ r,
+  deny /dev/bus/usb/ r,
+  deny /dev/bus/usb/001/ r,
+  deny /dev/bus/usb/002/ r,
```
Comment (roddhjav) [inline apparmor.d/groups/virt/containerd:59]:
```suggestion
  deny /dev/bus/usb/[0-9]*/ r,
```

File: apparmor.d/groups/virt/containerd [hunk 2]
```diff
@@ -40,5 +44,34 @@ profile containerd @{exec_path} {
...
+  deny /dev/char/ r,
+  deny /dev/cpu/ r,
+  deny /dev/cpu/0/ r,
+  deny /dev/cpu/1/ r,
```
Comment (roddhjav) [inline apparmor.d/groups/virt/containerd:63]:
```suggestion
  deny /dev/cpu/[0-9]*/ r,
```

File: apparmor.d/profiles-s-z/zpool [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  /dev/zfs rw,
+  @{PROC}/@{pids}/mounts r,
+
+  /dev/pts/[0-9]* rw,
+  /etc/hostid r,
+  @{PROC}/sys/kernel/spl/hostid r,
+
+  /run/blkid/blkid.tab wr,
+  /run/blkid/blkid.tab.old l,
+  /run/blkid/blkid.tab-* wrl,
```
Comment (roddhjav) [inline apparmor.d/profiles-s-z/zpool:25]:
```suggestion
  @{run}/blkid/blkid.tab wr,
  @{run}/blkid/blkid.tab.old l,
  @{run}/blkid/blkid.tab-* wrl,
```

**Comment by roddhjav** [comment]
Hi, thank a lot for your PR, I will merge it. However, they are a few things to change.

I commented your commits both in #52 and in #54 to give you a few example of what need to be changed. You should also have a look at the [profile guidelines](https://github.com/roddhjav/apparmor.d/blob/master/CONTRIBUTING.md#profile-guidelines) in order to be consistent in the structure of a profiles.

File: apparmor.d/profiles-s-z/zpool [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+@{exec_path} = /{usr/,}{local/,}{s,}bin/zpool
+profile zpool @{exec_path} flags=(complain) {
+  include <abstractions/base>
+  include <abstractions/disks-read>
+
+  capability sys_admin,
+
+  @{exec_path} rm,
+  /{usr/,}{local/,}lib/zfs-linux/zpool.d/* rix,
+  /{usr/,}{local/,}bin/{ba,da,k,z,}sh rix,
```
Comment (roddhjav) [inline apparmor.d/profiles-s-z/zpool:14]:
```suggestion
  /{usr/,}bin/{,ba,da}sh  rix,
```

File: apparmor.d/groups/virt/containerd [hunk 5]
```diff
@@ -40,5 +44,34 @@ profile containerd @{exec_path} {
   owner @{PROC}/@{pids}/mountinfo r,
         @{PROC}/sys/net/core/somaxconn r,
 
+  # Extracting container images
+  /usr/{local/,}bin/unpigz PUx,
+  
+  # zfs snapshotter
+  /{usr/,}{local/,}{s,}bin/zfs Px,
+  mount fstype=zfs -> /var/lib/containerd/tmpmounts/containerd-mount[0-9]*/,
+  umount -> /var/lib/containerd/tmpmounts/containerd-mount[0-9]*/,
```
Comment (roddhjav) [inline apparmor.d/groups/virt/containerd:53]:
```suggestion
  umount /var/lib/containerd/tmpmounts/containerd-mount[0-9]*/,
```

This fix https://github.com/roddhjav/apparmor.d/runs/7265316998

File: apparmor.d/groups/virt/containerd [hunk 6]
```diff
@@ -40,5 +44,34 @@ profile containerd @{exec_path} {
   owner @{PROC}/@{pids}/mountinfo r,
         @{PROC}/sys/net/core/somaxconn r,
 
+  # Extracting container images
+  /usr/{local/,}bin/unpigz PUx,
```
Comment (roddhjav) [inline apparmor.d/groups/virt/containerd:48]: Should be `/{usr/,}bin/unpigz rPUx,`

**Comment by Jeroen0494** [comment]
Thank you for the review, I really appreciate it!
I've applied all the aforementioned fixes.

**Comment by roddhjav** [comment]
That's almost good, again, have a look at the comments in both #52 and #54. Then, it will be ready to merge.

Sorry for the strong guideline regarding profile formatting, but it actually help a lot to share a common structure when you need to maintain **a lot** of profiles.

**Comment by Jeroen0494** [comment]
How's that? Not sure what else to change. I've put everything in the proper order specified in the contributing guideline.

**Comment by roddhjav** [comment]
It looks good now and ready to be merged!

**Comment by Jeroen0494** [comment]
Great! I don't have merge permissions, would you mind?

**Comment by roddhjav** [comment]
Merged! Thank

**Pull request merged**